### PR TITLE
RFC: Adds DuckDB support

### PR DESCRIPTION
**Pull request by cofin**
First off, thanks for the great library.  

I've added an adapter for `duckdb` and thought it might be helpful for others.  I've got 1 remaining test case to address, but it should be ready for consideration.

Just a few caveats I've seen so far:
- insert/update/delete returns `-1` instead of the updated row count
- Named parameters seem to have spotty support until 0.8 is released.

### CODE REVIEW

**Comment by cofin** [comment]
This is the cause for the insert returning issue: https://github.com/duckdb/duckdb/issues/6008

I've added a workaround and now all tests are passing.

File: aiosql/adapters/duckdb.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        cur = self._cursor(conn)
+        try:
+            cur.execute(sql, parameters)
+            result = cur.fetchone()
+            if result is not None and record_class is not None:
+                column_names = [c[0] for c in cur.description]
+                result = record_class(**dict(zip(column_names, result)))
+            elif result is not None and self.convert_row_to_dict:
+                column_names = [c[0] for c in cur.description]
+                result = dict(zip(column_names, result))
```
Comment (JacobCoffee) [inline aiosql/adapters/duckdb.py:60]:
```suggestion
                result = record_class(**dict(zip(column_names, result, strict=False)))
            elif result is not None and self.convert_row_to_dict:
                column_names = [c[0] for c in cur.description]
                result = dict(zip(column_names, result, strict=False))
```

File: aiosql/adapters/duckdb.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                        yield dict(zip(column_names, row))
+                    else:
+                        yield row
+            else:
+                first = True
+                for row in cur.fetchall():
+                    if first:  # only get description on the fly, for apsw
+                        column_names = [c[0] for c in cur.description]
+                        first = False
+                    yield record_class(**dict(zip(column_names, row)))
```
Comment (JacobCoffee) [inline aiosql/adapters/duckdb.py:46]:
```suggestion
                    yield record_class(**dict(zip(column_names, row, strict=False)))
```

File: aiosql/adapters/duckdb.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            cur.close()
+
+    def select_one(self, conn, _query_name, sql, parameters, record_class=None):
+        cur = self._cursor(conn)
+        try:
+            cur.execute(sql, parameters)
+            result = cur.fetchone()
+            if result is not None and record_class is not None:
+                column_names = [c[0] for c in cur.description]
+                result = record_class(**dict(zip(column_names, result)))
```
Comment (JacobCoffee) [inline aiosql/adapters/duckdb.py:57]:
```suggestion
                result = record_class(**dict(zip(column_names, result, strict=False)))
```

File: aiosql/adapters/duckdb.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            cur.execute(sql, parameters)
+            if record_class is None:
+
+                first = True
+                for row in cur.fetchall():
+                    if first:  # get column names on the fly
+                        column_names = [c[0] for c in cur.description]
+                        first = False
+                    if self.convert_row_to_dict:
+                        yield dict(zip(column_names, row))
```
Comment (JacobCoffee) [inline aiosql/adapters/duckdb.py:37]:
```suggestion
                        yield dict(zip(column_names, row, strict=False))
```

**Review by JacobCoffee — APPROVED** [review]
Not sure if you want to add zip if on 3.10+?

File: aiosql/adapters/duckdb.py (hunk 1 above)
Comment (cofin) [reply aiosql/adapters/duckdb.py]: I've updated.  Thanks for the review.

**Comment by zx80** [comment]
:+1: 

I'll look into it for the next release, planned in June after Python 3.7 support is dropped.

**Comment by zx80** [comment]
@cofin Maybe you could have a look at the docker stuff?

**Comment by cofin** [comment]
> @cofin Maybe you could have a look at the docker stuff?

Sure, I'd be happy to.  Which specific docker stuff are you referring to?

**Comment by zx80** [comment]
> Sure, I'd be happy to. Which specific docker stuff are you referring to?

The `docker` directory in the sources where there are docker files.

**Comment by zx80** [comment]
Merged in dev branch. Some more work to do though.

**Pull request closed without merging**